x/stake/endblock.go

```diff
@@ -252,3 +278,82 @@ func handleMatureUnbondingDelegations(k keeper.Keeper, ctx sdk.Context) ([]types
 
 	return completed, events
 }
+
+const (
+	maxProcessedRefundCount = 10
+)
+
+func handleRefundStake(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	iterator := k.IteratorAllDelegations(ctx)
+	defer iterator.Close()
+	var refundEvents sdk.Events
+	count := 0
+	bcValidators := k.GetAllValidators(ctx)
+	bcValidatorsMap := make(map[string]types.Validator, len(bcValidators))
+	for _, v := range bcValidators {
+		bcValidatorsMap[v.OperatorAddr.String()] = v
+	}
+	for ; iterator.Valid(); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.CDC(), iterator.Key(), iterator.Value())
+		if delegation.CrossStake {
+			result := handleMsgSideChainUndelegate(ctx.WithCrossStake(true), types.MsgSideChainUndelegate{
+				DelegatorAddr: delegation.DelegatorAddr,
+				ValidatorAddr: delegation.ValidatorAddr,
+				Amount:        sdk.NewCoin(k.BondDenom(ctx), delegation.GetShares().RawInt()),
+				SideChainId:   k.ScKeeper.BscSideChainId(ctx),
+			}, k)
+			refundEvents = refundEvents.AppendEvents(result.Events)
+		} else {
+			result := handleMsgUndelegate(ctx, types.MsgUndelegate{
```

Comment:
I think this is also `handleMsgSideChainUndelegate` but not `handleMsgUndelegate`, please confirm with Roshan

Reply:
fixed

Reply:
I think relayer fee is not calculate here, it is calculate when distribute reward in breath block.

And even if `delegation.CrossStake` is false, we need undelegate.

Reply:
why it comes back to `handleMsgUndelegate ` again? 😭

Reply:
fixed